src/app/qgshandlebadlayers.cpp

```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
+    for ( const QString altPath : altPaths )
+    {
+      if ( QFileInfo::exists( altPath + fileName ) && QFileInfo( altPath + fileName ).isFile() )
+        return ( altPath + fileName );
+    }
+  }
+  return ( mOriginalFileBase.value( layerId ) );
+}
+
+void QgsHandleBadLayers::autoFind()
+{
+  QDir::setCurrent( QgsProject::instance()->absolutePath() );
+  QgsProject::instance()->layerTreeRegistryBridge()->setEnabled( true );
+  buttonBox->button( QDialogButtonBox::Ignore )->setEnabled( false );
+
+  QList<int> layersToFind;
+  if ( mRows.size() > 0 )
+    layersToFind = mRows;
+  else
+  {
+    for ( int i = 0; i < mLayerList->rowCount(); i++ )
+      layersToFind.append( i );
+  }
+
+  const QList<int> constLayersToFind = layersToFind;
+  for ( int i : constLayersToFind )
+  {
+    int idx = mLayerList->item( i, 0 )->data( Qt::UserRole ).toInt();
+    QDomNode &node = const_cast<QDomNode &>( mLayers[ idx ] );
+
+    QTableWidgetItem *item = mLayerList->item( i, 4 );
+    QString datasource = item->text();
+    QString fileName;
+    const QString layerId { node.namedItem( QStringLiteral( "id" ) ).toElement().text() };
+    const QString name { mLayerList->item( i, 0 )->text() };
+    const QFileInfo dataInfo = QFileInfo( datasource );
+    const QString basepath = dataInfo.absoluteDir().path();
+    const QString longName = dataInfo.fileName();
+    QString provider = node.namedItem( QStringLiteral( "provider" ) ).toElement().text();
+    const QString fileType = mLayerList->item( i, 2 )->text();
+    if ( provider == "none" )
+    {
+      if ( mLayerList->item( i, 2 )->text() == "raster" )
+        provider = "gdal";
+      else if ( mLayerList->item( i, 2 )->text() == "vector" )
+        provider = "ogr";
+    }
+
+    QVariantMap providerMap = QgsProviderRegistry::instance()->decodeUri( dataInfo.absoluteFilePath(), provider );
+    if ( providerMap.contains( QStringLiteral( "path" ) ) )
+      fileName = QFileInfo( providerMap[ QStringLiteral( "path" ) ].toString() ).fileName();
+    else
+    {
+      item->setForeground( QBrush( Qt::red ) );
+      continue;
+    }
+
+    datasource = QDir::toNativeSeparators( checkBasepath( layerId, basepath, fileName ).replace( fileName, longName ) );
+
+    bool dataSourceChanged { false };
+
+    // Try first to change the datasource of the existing layers, this will
+    // maintain the current status (checked/unchecked) and group
+    if ( QgsProject::instance()->mapLayer( layerId ) )
+    {
+      QgsDataProvider::ProviderOptions options;
+      QgsMapLayer *mapLayer = QgsProject::instance()->mapLayer( layerId );
+      if ( mapLayer )
+      {
+        mapLayer->setDataSource( datasource, name, provider, options );
+        dataSourceChanged = mapLayer->isValid();
+      }
+    }
+
+    if ( !( dataSourceChanged ) )
+    {
+      datasource = QDir::toNativeSeparators( findFile( fileName, basepath ).replace( fileName, longName ) );
```

Comment:
again - better remove the toNativeSeparators here. (sorry if that was my fault!).

Reply:
Please see previous comment, as to why this is important.